### PR TITLE
[wip] adds rds deletion policy.

### PR DESCRIPTION
**Pull request by lsh-0**
I made the behaviour of 'buildercore.utils.ensure' sane again. The optional third parameter is now consistent with other projects also using ensure.
Tidied up 'buildercore.cfngen.build_context' a bit. It's still fat but we're getting there.

depends on:
* https://github.com/elifesciences/builder-base-formula/pull/99

### CODE REVIEW

File: src/buildercore/cfngen.py [hunk 1]
```diff
@@ -81,53 +74,37 @@ def from_existing_context(field, default_value):
...
-    bit_keys = ['project_name', 'instance_id', 'cluster_id']
-    bits = dict(zip(bit_keys, core.parse_stackname(stackname, all_bits=True)))
-    assert bits['project_name'] == pname, \
-        "the project name derived from the `stackname` doesn't match the given project name"
+    bits = core.parse_stackname(stackname, all_bits=True, idx=True)
+    ensure(bits['project_name'] == pname,
+           "the project name %r derived from the given `stackname` %r doesn't match" % (bits['project_name'], pname))
     context.update(bits)
 
+    # is this a production instance? if yes, then we'll do things like tweak the dns records ...
```
Comment (giorgiosironi) [inline src/buildercore/cfngen.py]: do we? The only usage I could find is https://github.com/elifesciences/builder-base-formula/blob/29588e1f92b461be39eab77fbf7e6b7241d8a856/elife/postfix.sls#L8 which we could drop in favor of `pillar.elife.env == 'prod'`. Reducing the size of the context can make it easier to observe and debug

File: src/buildercore/cfngen.py [hunk 2]
```diff
@@ -171,6 +145,43 @@ def _parameterize(string):
     return context
 
 
+def build_context_rds(context, existing_context):
```
Comment (giorgiosironi) [inline src/buildercore/cfngen.py]: seems a good way to express only RDS has a dependency on `existing_context`

File: src/buildercore/cfngen.py [hunk 3]
```diff
@@ -171,6 +145,43 @@ def _parameterize(string):
...
+    if 'rds' not in context['project']['aws']:
+        return {}
+    stackname = context['stackname']
+
+    # deletion policy
+    # enforce here rather than allow to be specified in project config
+    deletion_policy = 'Delete'
+    if context['is_prod_instance'] or context['instance_id'] in ['continuumtest', 'end2end']:
```
Comment (giorgiosironi) [inline src/buildercore/cfngen.py]: tricky to have these keys specified in keys rather than configuration :thinking:

Comment (giorgiosironi) [reply src/buildercore/cfngen.py]: I would prefer to go for safety and here have a default of `Snapshot`; in https://github.com/elifesciences/builder/blob/master/projects/elife.yaml#L165-L170 add a `rds` key alongside `ec2` with the override to `Delete`. It lets us treat adhoc environments (I just created `journal--annotationstest` that will be used for a month or so) as akin to production, while formula builds using `standalone` really are throwaway stacks.

**Comment by giorgiosironi** [comment]
I think `test_buildercore_trop` is the place to be, it should already have some RDS sample stacks.

File: src/buildercore/cfngen.py (hunk 2 above)
Comment (lsh-0) [reply src/buildercore/cfngen.py]: that's what I thought

File: src/buildercore/cfngen.py (hunk 3 above)
Comment (lsh-0) [reply src/buildercore/cfngen.py]: tweaking the logic so default is Snapshot is easy.

I originally had the deletion policy specified in the projects.yaml file and found myself repeating the line `deletion-policy: Delete` again and again and again. Repeating something that often is bound to cause a mistake. I prefer enforcing this in code, but there is room for an explicit override that need not be repeated. 

But yes - those keys are just magic strings floating there. I can reinforce them conf.py or core.py, but I don't seem to recall builder knowing too much about named environments. i'll investigate

File: src/buildercore/cfngen.py (hunk 1 above)
Comment (lsh-0) [reply src/buildercore/cfngen.py]: oh wow - that's some primordial Salt code :) I'll fix that formula up and remove this key.

**Comment by lsh-0** [comment]
did I add you as a reviewer or did you add yourself, @giorgiosironi ? I've been conscious of the fact that I don't want to impinge on another's time until a PR is ready, and that between opening a PR and getting a review more changes may happen. I could have sworn I deliberately left off adding reviewers for this

File: src/buildercore/cfngen.py (hunk 1 above)
Comment (lsh-0) [reply src/buildercore/cfngen.py]: pr: https://github.com/elifesciences/builder-base-formula/pull/99

File: src/buildercore/cfngen.py (hunk 3 above)
Comment (lsh-0) [reply src/buildercore/cfngen.py]: I'm grepping through builder's source and there is a distinct lack of code doing environment checking, which I like. I wish our formula's were this free of conditional behaviour. It's going to make the project files even more verbose, but I'll go with your recommendation: default to `Snapshot` and have the alt-configs explicitly set to `Delete`

**Comment by giorgiosironi** [comment]
I probably added myself because I am working in a similar area for `elasticache` and don't want to diverge

**Comment by lsh-0** [comment]
no worries. I'll merge this and then check your PR for any conflicts.

**Pull request merged**